src/expr/src/relation/mod.rs

```diff
@@ -363,24 +352,27 @@ impl RelationExpr {
                 //
                 // We are going to use the uniqueness constraints of the
                 // first relation, and attempt to use the presented order.
-                let remains_unique = (1..inputs.len()).all(|index| {
-                    let mut prior_bound = Vec::new();
-                    for equivalence in equivalences {
-                        if equivalence.iter().any(|expr| {
-                            expr.support()
-                                .into_iter()
-                                .all(|i| input_relation[i] < index)
-                        }) {
-                            for expr in equivalence {
-                                if let ScalarExpr::Column(c) = expr {
-                                    prior_bound.push(c);
+                let mut column_with_prior_bound_by_input = vec![HashSet::new(); inputs.len() - 1];
+                for equivalence in equivalences {
+                    let min_bound_input = equivalence
+                        .iter()
+                        .flat_map(|expr| input_mapper.lookup_inputs(expr).max())
+                        .min();
+                    if let Some(min_bound_input) = min_bound_input {
+                        for expr in equivalence {
+                            if let ScalarExpr::Column(c) = expr {
+                                let (col, input) = input_mapper.map_column_to_local(*c);
+                                if input > min_bound_input {
+                                    column_with_prior_bound_by_input[input - 1].insert(col);
                                 }
                             }
                         }
                     }
+                }
+                let remains_unique = (1..inputs.len()).all(|index| {
                     input_types[index].keys.iter().any(|ks| {
                         ks.iter()
-                            .all(|k| prior_bound.contains(&&(prior_arities[index] + k)))
+                            .all(|k| column_with_prior_bound_by_input[index - 1].contains(k))
                     })
                 });
```

Comment:
I'm tempted to say we should move all of this to the `JoinMapper` struct, as it seems to contain re-usable information about type structure. In particular, whatever analysis we want to use about how to find the unique keys of a join. There are other places in the code where we want to know about unique keys (e.g. in redundant join elision) and I suspect that we might want to just lift them out. If `RelationExpr` wrapped structs in each of its variants, we would move the logic to that struct for example.

Reply:
I think @justinj might have a dissenting opinion, in that key structure may be one of the things that should be un-baked from the "type".

Reply:
I've come to just think of what we call a "type" as "logical properties," so I have no strong feelings at this point about it